website/blog/2023-09-06-optimize-try-cast.mdx

```diff
@@ -0,0 +1,114 @@
+---
+slug: optimize-try_cast.
+title: "Learnings from optimizing try_cast"
+authors: [lsakka]
+tags: [tech-blog, performance]
+---
+
+One of the queries shadowed internally at Meta was much slower in Velox compared to presto(2 days vs 4.5h). Initial investigation identified that the overhead is related to casting empty strings inside a try_cast.
+
+In this blogpost I summarize my learnings from investigating and optimizing try_cast.
+
+--------------------------------------------------------------------------------------------
+
+### Start and end results
+
+Initial benchmark:
+```
+try_cast(empty_string_col as int)                     4.88s   204.83m
+try_cast(valid_string_col as int)                    2.15ms    465.22
+```
+
+The difference between casting a valid and invalid input is huge (>1000X), although ideally casting an invalid string shall be
+just setting a null and should not be that expensive.
+
+End results:
+```
+try_cast(empty_string_col as int)                    1.24ms    808.79
+try_cast(valid_string_col as int)                    2.15ms    465.22
+```
+--------------------------------------------------------------------------------------------
+
+### Sources of regressions
+The investigation revealed several factors that contributed to the huge gap, summarized in the points below in addition to
+their approximate significance.
+
+
+**Error logs overhead.**
+
+Whenever a VeloxUserError is thrown an error log used to be generated, however those errors are expected to, (1) either get converted to null if is
+thrown from within a try, (2) or show up to the user otherwise. Hence, no need for that expensive logging .
+
+Moreover, two logs used to be generated for each row since VELOX_USER_FAIL used to be called twice. Logging user errors was disabled and
+this translated to saving 2.6s of the 4.88s*.
+
+
+**Throwing overhead.**
+
+Each time a row is casted **four** exception used to be thrown:
+
+1. Folly library throw.
+2. cast function catch the exception and convert it to Velox exception and throw it in Conversions.h.
+3. CastToPrimitive catch the exception and throw it again with a different specific detailed message.
+4. Another throw happens inside applyToSelectedNoThrow, which used to catch that error and call setError which calls
```

Comment:
Finally, a forth throw came from applyToSelectedNoThrow which caught an exception and called toVeloxException to check exception type and re-throw.